dialer/proxyless.go

```diff
@@ -0,0 +1,97 @@
+package dialer
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/x/smart"
+)
+
+type proxylessDialer struct {
+	strategyFinder *smart.StrategyFinder
+	dialers        sync.Map
+}
+
+func newProxylessDialer() Dialer {
+	return &proxylessDialer{
+		strategyFinder: &smart.StrategyFinder{
+			TestTimeout:  5 * time.Second,
+			LogWriter:    nil,
+			StreamDialer: &transport.TCPDialer{},
+			PacketDialer: &transport.UDPDialer{},
+		},
+	}
+}
+
+// DialContext dials out to the domain or IP address representing a destination site.
+func (d *proxylessDialer) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	// Check if the dialer already exists in the map
+	if dialer, ok := d.dialers.Load(addr); ok {
+		// If it exists, use it
+		return dialer.(transport.StreamDialer).DialStream(ctx, addr)
+	}
+	// If it doesn't exist, create a new dialer
+	// Parse the domain from the address
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		host = addr
+	}
+	domains := []string{host}
+
+	configBytes, err := embedFS.ReadFile("smart_dialer_config.yml")
+	if err != nil {
+		log.Errorf("Failed to read smart dialer config: %v", err)
+		return nil, err
+	}
```

Comment:
We probably want to load this file once instead of reading every dial, so we could move this to `newProxylessDialer` and add a attribute for config on `proxylessDialer`

Reply:
@myleshorton what do you think about adding the smart dialer config as part of the global config? We could add different techniques later and it wouldn't be hard-coded

Reply:
I'm afraid that users in the future might stop updating on this version (by any random reason which happens) and the techniques implemented here would be frozen